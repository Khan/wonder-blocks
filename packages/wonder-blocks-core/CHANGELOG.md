# @khanacademy/wonder-blocks-core

## 5.4.0

### Minor Changes

-   1344436f: Add `PropsFor` type to simplify `JSX.LibraryManagedAttributes` usage

## 5.3.1

### Patch Changes

-   9f3752d4: Used named functions in componenets with forwarded refs

## 5.3.0

### Minor Changes

-   c37b99aa: Forward refs for Text and Title
-   c4cef3e6: Added ref forwarding to View

## 5.2.3

### Patch Changes

-   d4c412b5: Update useForceUpdate to ensure consuming hooks properly refresh

## 5.2.2

### Patch Changes

-   64a188e3: Fix useForceUpdate so that regardless of how many times it is called before a new render, it will always cause a new render

## 5.2.1

### Patch Changes

-   5a1ea891: Make sure that React contexts are named
-   df9a10aa: Update state and props to be readonly in components using getDerivedStateFromProps()

## 5.2.0

### Minor Changes

-   fa70c895: Add 'lang', 'className' and 'htmlFor' properties to Text and View

### Patch Changes

-   19ab0408: Update flowgen to convert React.ForwardRefExoticComponent<> and React.FowardedRef<> properly
-   fa70c895: Remove all TypeScript error suppressions on JSX attributes

## 5.1.0

### Minor Changes

-   3c400719: Add useLatestRef hook to wonder-blocks-core

### Patch Changes

-   a6164ed0: Don't use React.FC<> for functional components

## 5.0.4

### Patch Changes

-   c20f48f3: Don't transpile classes when building bundles

## 5.0.3

### Patch Changes

-   b281e2eb: Fix generate Flow types for real

## 5.0.2

### Patch Changes

-   21ce20c7: Fix issues with generate flow types in wonder-blocks-core

## 5.0.1

### Patch Changes

-   ccb6fe00: Miscellaneous TS type fixes
-   d4c2b18c: Fix a variety of issues with Flow types generated by flowgen

## 5.0.0

### Major Changes

-   1ca4d7e3: Fix minor issue with generate Flow types (this is a major bump b/c I forgot to do one after doing the TS conversion)

## 4.9.1

### Patch Changes

-   b5ba5568: Ensure that flow lib defs use React.ElementConfig<> isntead of JSX.LibraryManagedAttributes<>

## 4.9.0

### Minor Changes

-   779b031d: Update AriaProps to use React's AriaAttributes and AriaRole types

## 4.8.0

### Minor Changes

-   873f4a14: Update aphrodite lib def and StyleType

### Patch Changes

-   d816af08: Update build and test configs use TypeScript
-   3891f544: Update babel config to include plugins that Storybook needed
-   0d28bb1c: Configured TypeScript
-   3d05f764: Fix HOCs and other type errors
-   c2ec4902: Update eslint configuration, fix lint
-   2983c05b: Include 'types' field in package.json
-   77ff6a66: Generate Flow types from TypeScript types
-   ec8d4b7f: Fix miscellaneous TypeScript errors

## 4.7.0

### Minor Changes

-   91cb727c: Convert enums to POJOs

### Patch Changes

-   91cb727c: Remove file extensions from imports

## 4.6.2

### Patch Changes

-   496119f2: Cleanup WB interdependencies

## 4.6.1

## 4.6.0

### Minor Changes

-   b561425a: Add useIsMounted() hook
-   a566e232: Add useOnMountEffect hook
-   d2b21a6e: Export useOnMountEffect hook

## 4.5.0

### Minor Changes

-   175a2dd2: Add 'useRenderState' hook

## 4.4.0

### Minor Changes

-   6ee20af9: Add `useOnline` hook to encapsulate navigator.onLine and the offline/online events

## 4.3.2

### Patch Changes

-   5f4a4297: Make dependency on `flow-enums-runtime` explicit
-   2b96fd59: Change flow-enums-runtime to be peer dependencies

## 4.3.1

## 4.3.0

### Minor Changes

-   246a921d: NEW: `useForceUpdate` hook. This should rarely be used and likely only ever from other hooks.

## 4.2.1

## 4.2.0

### Minor Changes

-   0aed8723: Adds `throwIfNested` required prop to `RenderStateRoot`.

## 4.1.0

### Minor Changes

-   45588e5f: Fix an issue with `useUniqueIdWithMock`/`useUniqueIdWithoutMock`
    rerender more than was needed. The fix introduces `<RenderStateRoot>`
    which must be an ancestor to all components uses these hooks.
-   875b7893: Nesting of `RenderStateRoot`s inside each other can result in extra renders
    and potentially incorrect behavior. `RenderStateRoot` now throws if it
    appears as a descendent of another `RenderStateRoot`.
