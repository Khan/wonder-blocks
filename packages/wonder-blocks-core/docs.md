`wonder-blocks-core` provides building blocks for other `wonder-blocks` 
components.  These components, `View` and `Text` can also be used on their 
own.

`View` roughly maps to `div` and `Text` roughly maps to `span`.

```js
const {StyleSheet} = require("aphrodite");

const styles = StyleSheet.create({
    container: {
        padding: 32,
        backgroundColor: "plum",
    },
    text: {
        fontFamily: "sans-serif",
        fontSize: 24,
    },
});

<View style={styles.container}>
    <Text style={styles.text}>
        Hello, world!
    </Text>
</View>
```

Other props can be passed through `View` or `Text`, as if they were normal tags.

```js
<View>
    <View onClick={() => alert("Clicked!")}>
        Click me!
    </View>

    <Text aria-hidden>
        This text is hidden from screen readers.
    </Text>
</View>
```

There's also an HOC that adds the same style prop to existing 
components.

```js
const {addStyle} = require("./index.js");
const {StyleSheet} = require("aphrodite");

const styles = StyleSheet.create({
    // default style for all instances of StyledInput
    input: {
        fontSize: 30,
    },
    // style for a particular instance of StyledInput
    pink: {
        backgroundColor: "pink",
    },
});

const StyledInput = addStyle("input", styles.input);

<StyledInput style={styles.pink} type="text" placeholder="hello, world"/>;
```
