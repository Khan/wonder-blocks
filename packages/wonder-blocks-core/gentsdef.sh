#!/usr/bin/env bash
# TODO: figure out why writing this as script in package.json doesn't work
# My hunch is that it has to do with where tsconfig.json is located
tsc --emitDeclarationOnly --declaration --declarationDir dist --allowSyntheticDefaultImports --downlevelIteration --jsx react --lib ES2018,DOM index.ts
