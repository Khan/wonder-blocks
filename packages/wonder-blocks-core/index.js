// @flow
import type {
    ClickableHandlers,
    ClickableState,
} from "./components/clickable-behavior.js";
import type {AriaProps, IIdentifierFactory, StyleType} from "./util/types.js";
import type {Intersection} from "./util/get-element-intersection.js";

export {default as ClickableBehavior} from "./components/clickable-behavior.js";
export {default as Text} from "./components/text.js";
export {default as View} from "./components/view.js";
export {
    default as WithSSRPlaceholder,
} from "./components/with-ssr-placeholder.js";
export {default as IDProvider} from "./components/id-provider.js";
export {default as UniqueIDProvider} from "./components/unique-id-provider.js";
export {default as addStyle} from "./util/add-style.js";
export {
    default as getClickableBehavior,
} from "./util/get-clickable-behavior.js";
export {
    default as getElementIntersection,
} from "./util/get-element-intersection.js";

export type {
    AriaProps,
    ClickableHandlers,
    ClickableState,
    Intersection,
    IIdentifierFactory,
    StyleType,
};
