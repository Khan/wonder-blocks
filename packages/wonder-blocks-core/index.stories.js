// @flow

import React from "react";
import {storiesOf} from "@storybook/react";
import {StyleSheet} from "aphrodite";

import {Text, View} from "./index.js";

storiesOf("Core", module).addWithJSX("all", () => (
    <div>
        <Text>Text</Text>
        <View>View</View>
    </div>
));

storiesOf("Core/Text", module)
    .addWithJSX("text", () => <Text>Text</Text>)
    .addWithJSX("aphrodite styles", () => <Text style={styles.pink}>Text</Text>)
    .addWithJSX("multiple Aphrodite styles", () => (
        <Text style={[styles.pink, styles.bold]}>Text</Text>
    ))
    .addWithJSX("multiple text elements", () => (
        <div>
            <Text style={styles.bold}>Text</Text>
            <Text style={styles.pink}>Text</Text>
        </div>
    ));

storiesOf("Core/View", module)
    .addWithJSX("view", () => <View>View</View>)
    .addWithJSX("aphrodite styles", () => <View style={styles.pink}>View</View>)
    .addWithJSX("multiple Aphrodite styles", () => (
        <View style={[styles.pink, styles.bold]}>Text</View>
    ))
    .addWithJSX("multiple view elements", () => (
        <div>
            <View style={styles.bold}>Text</View>
            <View style={styles.pink}>Text</View>
        </div>
    ));

const styles = StyleSheet.create({
    pink: {
        backgroundColor: "pink",
    },
    bold: {
        fontWeight: "bold",
    },
});
