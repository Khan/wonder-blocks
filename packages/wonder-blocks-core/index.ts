// @flow
import {ClickableHandlers} from "./components/clickable-behavior";
import {AriaProps, IIdentifierFactory, StyleType} from "./util/types";
import {Intersection} from "./util/get-element-intersection";

export {default as ClickableBehavior} from "./components/clickable-behavior";
export {default as Text} from "./components/text";
export {default as View} from "./components/view";
export {
    default as WithSSRPlaceholder,
} from "./components/with-ssr-placeholder";
export {default as UniqueIDProvider} from "./components/unique-id-provider";
export {default as addStyle} from "./util/add-style";
export {
    default as getClickableBehavior,
} from "./util/get-clickable-behavior";
export {
    default as getElementIntersection,
} from "./util/get-element-intersection";

export {
    AriaProps,
    ClickableHandlers,
    Intersection,
    IIdentifierFactory,
    StyleType,
};
