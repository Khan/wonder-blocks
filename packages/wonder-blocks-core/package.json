{
  "name": "@khanacademy/wonder-blocks-core",
  "version": "1.1.1",
  "design": "v1",
  "publishConfig": {
    "access": "public"
  },
  "description": "",
  "main": "dist/index.js",
  "source": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "@khanacademy/wonder-blocks-spacing": "^2.0.4"
  },
  "peerDependencies": {
    "aphrodite": "^1.2.5",
    "prop-types": "^15.6.2",
    "react": "^16.4.1",
    "react-dom": "^16.4.1",
    "react-router-dom": "^4.2.2"
  },
  "author": "",
  "license": "MIT"
}
