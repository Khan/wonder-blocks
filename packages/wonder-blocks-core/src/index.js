// @flow
import type {AriaProps, IIdentifierFactory, StyleType} from "./util/types.js";
import type {Intersection} from "./util/get-element-intersection.js";

export {default as Text} from "./components/text.js";
export {default as View} from "./components/view.js";
export {default as WithSSRPlaceholder} from "./components/with-ssr-placeholder.js";
export {default as IDProvider} from "./components/id-provider.js";
export {default as UniqueIDProvider} from "./components/unique-id-provider.js";
export {default as addStyle} from "./util/add-style.js";
export {default as getElementIntersection} from "./util/get-element-intersection.js";
export {default as Server} from "./util/server.js";
export {useRenderState} from "./hooks/use-render-state.js";
export {
    useUniqueIdWithMock,
    useUniqueIdWithoutMock,
} from "./hooks/use-unique-id.js";

export type {AriaProps, Intersection, IIdentifierFactory, StyleType};
