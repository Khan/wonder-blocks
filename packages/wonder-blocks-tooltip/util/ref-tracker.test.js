// @flow
import * as React from "react";
import * as ReactDOM from "react-dom";
import {mount} from "enzyme";

import {View} from "@khanacademy/wonder-blocks-core";
import RefTracker from "./ref-tracker.js";

type CallbackFn = (?HTMLElement) => void;

describe("RefTracker", () => {
    describe("#setCallback", () => {
        test("called with falsy value, no throw", () => {
            // Arrange
            const tracker = new RefTracker();

            // Act
            const underTest = () => tracker.setCallback(null);

            // Assert
            expect(underTest).not.toThrow();
        });

        test("called with non-function, throws", () => {
            // Arrange
            const tracker = new RefTracker();
            const targetFn = (({}: any): CallbackFn);

            // Act
            const underTest = () => tracker.setCallback(targetFn);

            // Assert
            expect(underTest).toThrowErrorMatchingSnapshot();
        });

        describe("called with a function", () => {
            test("no prior call to updateRef, does not call targetFn", () => {
                // Arrange
                const tracker = new RefTracker();
                const targetFn = jest.fn();
                // Act
                tracker.setCallback(targetFn);

                // Assert
                expect(targetFn).not.toHaveBeenCalled();
            });

            test("prior updateRef call, target called with ref's node", () => {
                // Arrange
                const tracker = new RefTracker();
                const targetFn = jest.fn();
                const arrange = (actAssert) => {
                    const nodes = (
                        <View>
                            <View ref={actAssert} />
                        </View>
                    );
                    mount(nodes);
                };

                const actAndAssert = (ref) => {
                    const domNode = ReactDOM.findDOMNode(ref);
                    tracker.updateRef(ref);

                    // Act
                    tracker.setCallback(targetFn);

                    // Assert
                    expect(targetFn).toHaveBeenCalledWith(domNode);
                };

                arrange(actAndAssert);
            });
        });
    });

    describe("#updateRef", () => {
        describe("calling without setting a callback", () => {
            test("falsy ref, no throw", () => {
                // Arrange
                const tracker = new RefTracker();

                // Act
                const underTest = () => tracker.updateRef(null);

                // Assert
                expect(underTest).not.toThrow();
            });

            test("real ref, no callback, no throw", () => {
                // Arrange
                const tracker = new RefTracker();
                const arrange = (actAssert) => {
                    const nodes = (
                        <View>
                            <View ref={actAssert} />
                        </View>
                    );
                    mount(nodes);
                };

                const actAndAssert = (ref) => {
                    // Act
                    const underTest = () => tracker.updateRef(ref);

                    // Assert
                    expect(underTest).not.toThrow();
                };

                arrange(actAndAssert);
            });

            test("real ref, targetFn callback, calls the targetFn", () => {
                // Arrange
                const tracker = new RefTracker();
                const targetFn = jest.fn();
                tracker.setCallback(targetFn);

                const arrange = (actAssert) => {
                    const nodes = (
                        <View>
                            <View ref={actAssert} />
                        </View>
                    );
                    mount(nodes);
                };

                const actAndAssert = (ref) => {
                    const domNode = ReactDOM.findDOMNode(ref);

                    // Act
                    tracker.updateRef(ref);

                    // Assert
                    expect(targetFn).toHaveBeenCalledWith(domNode);
                };

                arrange(actAndAssert);
            });

            test("same ref, targetFn callback, does not call targetFn", () => {
                // Arrange
                const tracker = new RefTracker();
                const targetFn = jest.fn();
                tracker.setCallback(targetFn);

                const arrange = (actAssert) => {
                    const nodes = (
                        <View>
                            <View ref={actAssert} />
                        </View>
                    );
                    mount(nodes);
                };

                const actAndAssert = (ref) => {
                    tracker.updateRef(ref);
                    targetFn.mockClear();

                    // Act
                    tracker.updateRef(ref);

                    // Assert
                    expect(targetFn).not.toHaveBeenCalled();
                };

                arrange(actAndAssert);
            });
        });
    });
});
