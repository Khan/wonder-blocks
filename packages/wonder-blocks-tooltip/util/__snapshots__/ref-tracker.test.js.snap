// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefTracker #setCallback called with non-function, throws 1`] = `"targetFn must be a function"`;
